File: tests/api_loading/consumers/ProductionProbe.swift
import Foundation
import loading_native

private final class ProductionSink: loading_native.ModalitySink, @unchecked Sendable {
  let session: loading_native.Session
  private let lock = NSLock()
  private var completed: [loading_native.FinishReason] = []
  private var content = ""
  var done: [loading_native.FinishReason] { lock.withLock { completed } }
  var text: String { lock.withLock { content } }

  init(_ session: loading_native.Session) { self.session = session }
  func onThoughtChunk(text: String) { preconditionFailure("unexpected thoughts") }
  func onAudioFrames(pcm: [Float], sampleRate: UInt32) {
    preconditionFailure("unexpected audio")
  }
  func onTextChunk(text: String) {
    lock.withLock { content += text }
    session.cancel()
  }
  func onDone(reason: loading_native.FinishReason) { lock.withLock { completed.append(reason) } }
}

func productionConsumed(_ loader: loading_native.ModelLoader) {
  for direct in [false, true] {
    do {
      if direct { _ = try loader.buildGenerative() } else { _ = try loader.build() }
      preconditionFailure("production loader reused")
    } catch loading_native.LoadError.Consumed {} catch {
      preconditionFailure("unexpected consumed error: \(error)")
    }
  }
}

func runProduction(_ root: URL, bytes: Data) throws -> [String] {
  for (index, backend) in [BackendPreference.auto, .cpu, .gpu, .metal, .hexagon].enumerated() {
    var encoded: [UInt8] = []
    FfiConverterTypeBackendPreference.write(backend, into: &encoded)
    precondition(encoded == [0, 0, 0, UInt8(index + 1)])
    var buffer = (data: Data(encoded), offset: 0)
    let decoded = try FfiConverterTypeBackendPreference.read(from: &buffer)
    precondition(decoded == backend && buffer.offset == 4)
  }
  for tag: Int32 in [0, 6, -1, Int32.max] {
    var wire = tag.bigEndian
    let data = withUnsafeBytes(of: &wire) { Data($0) }
    var buffer = (data: data, offset: 0)
    do {
      _ = try FfiConverterTypeBackendPreference.read(from: &buffer)
      preconditionFailure("malformed production backend enum accepted")
    } catch {
      precondition(error.localizedDescription == "Raw enum value doesn't match any cases")
      precondition(buffer.offset == 4)
    }
  }
  for backend in [BackendPreference.gpu, .metal] {
    for direct in [false, true] {
      let loader = loading_native.ModelLoader(
        source: .bytes(bytes: bytes), config: EngineConfig(backend: backend))
      do {
        if direct { _ = try loader.buildGenerative() } else { _ = try loader.build() }
        preconditionFailure("unavailable production backend accepted")
      } catch loading_native.LoadError.Assembly(let actual, let detail) {
        precondition(actual == (backend == .gpu ? "Gpu" : "Metal") && !detail.isEmpty)
      }
      productionConsumed(loader)
    }
  }
  let defaults = loading_native.EngineConfig(backend: .cpu)
  precondition(defaults.contextSize == 4096 && defaults.bundleRepo == nil)
  precondition(defaults.draftModel == nil && !defaults.gpuDepthformer)
  let sessionDefaults = loading_native.SessionConfig()
  precondition(sessionDefaults.maxSeqLen == nil && sessionDefaults.kvCompression == nil)
  precondition(sessionDefaults.nKeep == 0 && sessionDefaults.seed == nil)
  precondition(sessionDefaults.ubatchSize == 512 && !sessionDefaults.gpuDepthformer)
  let path = root.appendingPathComponent("model.gguf").path
  let sources: [loading_native.ModelSource] = [
    .bytes(bytes: bytes), .path(path: path),
    .files(
      files: loading_native.ModelFiles(
        model: path, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil,
        draftModel: nil, extras: [:], inferenceType: nil, chatTemplate: nil)),
    .parts(
      parts: loading_native.ModelParts(
        model: bytes, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil,
        draftModel: nil, inferenceType: nil, chatTemplate: nil, generationDefaults: nil)),
  ]
  for source in sources {
    for direct in [false, true] {
      var loader: loading_native.ModelLoader? = loading_native.ModelLoader(
        source: source, config: defaults)
      var handle: loading_native.ModelHandle? = direct ? nil : try loader!.build()
      var model: loading_native.GenerativeModel? =
        direct ? try loader!.buildGenerative() : handle!.asGenerative()
      productionConsumed(loader!)
      var engine: loading_native.CeraEngine? = model!.engine()
      precondition(enginesShareForProbe(first: engine!, second: model!.engine()))
      precondition(engine!.contextSize() == 4096 && engine!.metadata().maxSeqLen == 64)
      let independent = try loading_native.CeraEngine.fromBytes(bytes: bytes, config: defaults)
      precondition(!enginesShareForProbe(first: engine!, second: independent))
      let config = loading_native.SessionConfig(
        maxSeqLen: 8, kvCompression: .f16, seed: UInt64.max, ubatchSize: 1,
        gpuDepthformer: true)
      let session: loading_native.Session = try model!.createSession(config: config)
      let observed = try sessionConfigForProbe(session: session)
      precondition(observed == config)
      let sibling = try engine!.newSession(config: loading_native.SessionConfig())
      weak let releasedLoader = loader
      weak let releasedHandle = handle
      weak let releasedModel = model
      loader = nil
      handle = nil
      model = nil
      precondition(releasedLoader == nil && releasedHandle == nil && releasedModel == nil)
      precondition(engine!.contextSize() == 4096)
      precondition(engine!.encodeText(text: "ab") == [0, 1])
      precondition(engine!.decodeTokens(tokens: [0, 1]) == "ab")
      try sibling.appendTokens(tokens: [1])
      try session.appendTokens(tokens: [0, 1])
      engine!.clearPrefixCache()
      precondition(session.position() == 2 && sibling.position() == 1)
      weak let releasedEngine = engine
      engine = nil
      precondition(releasedEngine == nil)
      let one = loading_native.GenerateOpts(maxTokens: 1, temperature: 0.7, ignoreEos: true)
      let first = try session.generate(opts: one)
      precondition(first.tokens.count == 1 && session.position() == 3)
      let two = loading_native.GenerateOpts(maxTokens: 2, temperature: 0.7, ignoreEos: true)
      let next = try session.generate(opts: two)
      precondition(next.tokens.count == 2 && session.position() == 5)
      precondition(sibling.position() == 1)
      try session.reset()
      precondition(session.position() == 0)
      do {
        try session.appendTokens(tokens: [])
        preconditionFailure("empty input accepted")
      } catch loading_native.FfiError.EmptyInput {}
      try session.appendTokens(tokens: [0, 1])
      let sink = ProductionSink(session)
      let streamed = try session.generateStreaming(
        opts: loading_native.GenerateOpts(
          maxTokens: 3, temperature: 0.7, ignoreEos: true, flushEveryTokens: 1), sink: sink)
      precondition(!sink.text.isEmpty && sink.done.count == 1)
      precondition(streamed.finishReason == .cancelled && streamed.tokensGenerated == 1)
      let position = session.position()
      session.clearCancel()
      _ = try session.generate(opts: one)
      precondition(session.position() == position + 1)
    }
  }
  let loader = loading_native.ModelLoader(source: .bytes(bytes: bytes), config: defaults)
  let model = try loader.buildGenerative()
  let modes: [loading_native.KvCompression?] = [
    nil, loading_native.KvCompression.none, .f16,
    .turboQuant(seed: 0, keys: true, values: true),
    .turboQuant(seed: UInt64.max, keys: true, values: false),
    .turboQuant(seed: 42, keys: false, values: true),
  ]
  for mode in modes {
    let config = loading_native.SessionConfig(
      maxSeqLen: 8, kvCompression: mode, nKeep: 1,
      seed: 0, ubatchSize: 0, gpuDepthformer: true)
    let session = try model.createSession(config: config)
    var expected = config
    expected.kvCompression = mode ?? loading_native.KvCompression.none
    let observed = try sessionConfigForProbe(session: session)
    precondition(observed == expected)
    try session.appendTokens(tokens: [0, 1])
    let generated = try session.generate(
      opts: loading_native.GenerateOpts(
        maxTokens: 1, temperature: 0, ignoreEos: true))
    precondition(generated.tokens.count == 1)
  }
  let capped = try model.createSession(config: loading_native.SessionConfig(maxSeqLen: 1))
  try capped.appendTokens(tokens: [0])
  do {
    try capped.appendTokens(tokens: [1])
    preconditionFailure("session cap ignored")
  } catch let loading_native.FfiError.ContextOverflow(maxSeqLen, by) {
    precondition(maxSeqLen == 1 && by == 1 && capped.position() == 1)
  }
  return try runProductionDefaults(bytes) + runProductionErrors(root, bytes) + [
    "production-backend-transport",
    "production-defaults", "production-sources", "production-session",
    "production-shared-engine", "production-kv-config", "production-stream-cancel",
  ]
}

func runProductionDefaults(_ bytes: Data) throws -> [String] {
  let empty = loading_native.SamplingDefaults(
    temperature: nil, topP: nil, topK: nil, minP: nil, repetitionPenalty: nil)
  let sampling = loading_native.SamplingDefaults(
    temperature: 0.37, topP: 0.71, topK: 7, minP: 0.13, repetitionPenalty: 1.23)
  var profiles: [(String, loading_native.GenerationDefaults?, loading_native.GenerationDefaults)] =
    [
      ("absent", nil, .text(sampling: empty)),
      ("text-empty", .text(sampling: empty), .text(sampling: empty)),
    ]
  for (name, value) in [
    (
      "audio",
      loading_native.GenerationDefaults.audio(
        sampling: sampling, numberOfDecodingThreads: 3, audioTemperature: 0.625, audioTopK: 11)
    ),
    (
      "audio",
      .audio(
        sampling: empty, numberOfDecodingThreads: 0, audioTemperature: 0, audioTopK: 0)
    ),
    (
      "audio",
      .audio(
        sampling: empty, numberOfDecodingThreads: UInt32.max, audioTemperature: 1,
        audioTopK: UInt32.max)
    ),
    (
      "audio-empty",
      .audio(
        sampling: empty, numberOfDecodingThreads: nil, audioTemperature: nil, audioTopK: nil)
    ),
  ] {
    profiles.append((name, value, value))
  }
  for (raw, canonical) in [
    (
      " { \"nested\" : [true, null, {\"x\":7}], \"label\": \"line\\ntext\" } ",
      "{\"label\":\"line\\ntext\",\"nested\":[true,null,{\"x\":7}]}"
    ),
    ("[1, 2, null]", "[1,2,null]"), ("42", "42"), ("true", "true"),
    ("\"text\"", "\"text\""), ("null", "null"),
  ] {
    profiles.append(("other", .other(rawJson: raw), .other(rawJson: canonical)))
  }
  for (_, defaults, expected) in profiles {
    for typed in [false, true] {
      let parts = loading_native.ModelParts(
        model: bytes, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil,
        draftModel: nil, inferenceType: nil, chatTemplate: nil, generationDefaults: defaults)
      var loader: loading_native.ModelLoader? = loading_native.ModelLoader(
        source: .parts(parts: parts), config: EngineConfig(backend: .cpu))
      var handle: loading_native.ModelHandle?
      var model: loading_native.GenerativeModel?
      if typed {
        model = try loader!.buildGenerative()
      } else {
        handle = try loader!.build()
        model = handle!.asGenerative()
      }
      productionConsumed(loader!)
      let observed = productionDefaultsForProbe(engine: model!.engine())
      let session = try model!.createSession(config: SessionConfig(seed: 42))
      weak let releasedLoader = loader
      weak let releasedHandle = handle
      weak let releasedModel = model
      loader = nil
      handle = nil
      model = nil
      precondition(releasedLoader == nil && releasedHandle == nil && releasedModel == nil)
      precondition(observed == expected)
      try session.appendTokens(tokens: [0, 1])
      let result = try session.generate(
        opts: GenerateOpts(maxTokens: 3, temperature: 0, ignoreEos: true))
      precondition(result.tokens == [0, 1, 0] && session.position() == 5)
    }
  }
  for raw in ["", "{", "null trailing", "{\"x\":NaN}"] {
    for typed in [false, true] {
      let parts = loading_native.ModelParts(
        model: bytes, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil,
        draftModel: nil, inferenceType: nil, chatTemplate: nil,
        generationDefaults: .other(rawJson: raw))
      let loader = loading_native.ModelLoader(
        source: .parts(parts: parts), config: EngineConfig(backend: .cpu))
      do {
        if typed { _ = try loader.buildGenerative() } else { _ = try loader.build() }
        fatalError("malformed defaults succeeded")
      } catch let loading_native.LoadError.InvalidConfig(field, value, reason, detail) {
        precondition(field == "generation_defaults.raw_json" && value == raw)
        precondition(reason == "invalid_json" && !detail.isEmpty)
      }
      productionConsumed(loader)
    }
  }
  return Set(profiles.map { "production-parts-defaults-\($0.0)" }).sorted()
    + ["production-parts-defaults-invalid-json"]
}

private func runProductionErrors(_ root: URL, _ bytes: Data) throws -> [String] {
  var cases: [String] = []
  let parts = loading_native.ModelParts(
    model: bytes, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil, draftModel: nil,
    inferenceType: nil, chatTemplate: nil, generationDefaults: nil)
  for (arch, kind) in [
    ("bert", "Encoder"), ("modernbert", "Encoder"), ("whisper", "Whisper"), ("silero_vad", "Vad"),
    ("kws", "Hotword"),
  ] {
    for typed in [false, true] {
      let data = try Data(contentsOf: root.appendingPathComponent("\(arch).gguf"))
      let bad = loading_native.ModelLoader(
        source: .bytes(bytes: data), config: EngineConfig(backend: .metal))
      do {
        if typed { _ = try bad.buildGenerative() } else { _ = try bad.build() }
        fatalError("wrong kind succeeded")
      } catch let loading_native.LoadError.KindMismatch(expected, actual, architecture) {
        precondition(
          expected == "Generative" && actual == kind && architecture == arch,
          "\(arch) kind mismatch payload (typed=\(typed))")
      }
      productionConsumed(bad)
    }
    cases.append("production-kind-\(arch)")
  }
  for name in ["unknown", "malformed", "backend", "assembly", "inference"] {
    for typed in [false, true] {
      let data =
        name == "unknown"
        ? try Data(contentsOf: root.appendingPathComponent("unknown.gguf"))
        : name == "assembly"
          ? try Data(contentsOf: root.appendingPathComponent("llama.gguf"))
          : name == "malformed" ? Data([0, 1]) : bytes
      var unsupportedParts = parts
      unsupportedParts.inferenceType = "future/unsupported"
      let bad = loading_native.ModelLoader(
        source: name == "inference" ? .parts(parts: unsupportedParts) : .bytes(bytes: data),
        config: EngineConfig(backend: name == "backend" ? .metal : .cpu))
      do {
        if typed { _ = try bad.buildGenerative() } else { _ = try bad.build() }
        fatalError("invalid load succeeded")
      } catch let loading_native.LoadError.UnsupportedArchitecture(architecture) {
        precondition(name == "unknown" && architecture == "future_probe")
      } catch let loading_native.LoadError.Source(sourceKind, detail) {
        precondition(name == "malformed" && sourceKind == "bytes" && !detail.isEmpty)
      } catch let loading_native.LoadError.Assembly(backend, detail) {
        precondition(name == "backend" || name == "assembly")
        precondition(backend == (name == "backend" ? "Metal" : "Cpu") && !detail.isEmpty)
      } catch let loading_native.LoadError.UnsupportedInferenceType(inferenceType) {
        precondition(name == "inference" && inferenceType == "future/unsupported")
      }
      productionConsumed(bad)
    }
    cases.append("production-" + name)
  }
  return cases
}

File: tests/api_loading/consumers/ProductionProbe.kt
package loadingprobe

import uniffi.cera_ffi.BackendPreference
import uniffi.cera_ffi.CeraEngine
import uniffi.cera_ffi.EngineConfig
import uniffi.cera_ffi.FfiConverterTypeBackendPreference
import uniffi.cera_ffi.FfiException
import uniffi.cera_ffi.FinishReason
import uniffi.cera_ffi.GenerateOpts
import uniffi.cera_ffi.GenerationDefaults
import uniffi.cera_ffi.KvCompression
import uniffi.cera_ffi.LoadException
import uniffi.cera_ffi.ModalitySink
import uniffi.cera_ffi.ModelFiles
import uniffi.cera_ffi.ModelLoader
import uniffi.cera_ffi.ModelParts
import uniffi.cera_ffi.SamplingDefaults
import uniffi.cera_ffi.Session
import uniffi.cera_ffi.SessionConfig
import uniffi.loading_native.enginesShareForProbe
import uniffi.loading_native.productionDefaultsForProbe
import uniffi.loading_native.sessionConfigForProbe
import java.io.File
import java.nio.ByteBuffer
import uniffi.cera_ffi.ModelSource as Source

private class ProductionSink(
    val session: Session,
) : ModalitySink {
    val done = mutableListOf<FinishReason>()
    var text = ""

    override fun onThoughtChunk(text: String) {
        error("unexpected thoughts")
    }

    override fun onAudioFrames(
        pcm: List<Float>,
        sampleRate: UInt,
    ) {
        error("unexpected audio")
    }

    override fun onTextChunk(text: String) {
        this.text += text
        session.cancel()
    }

    override fun onDone(reason: FinishReason) {
        done += reason
    }
}

fun productionConsumed(loader: ModelLoader) {
    for (direct in listOf(false, true)) {
        try {
            if (direct) loader.buildGenerative() else loader.build()
            error("production loader reused")
        } catch (_: uniffi.cera_ffi.LoadException.Consumed) {
        }
    }
}

fun runProduction(
    root: File,
    bytes: ByteArray,
): List<String> {
    for ((index, backend) in BackendPreference.entries.withIndex()) {
        val buffer = ByteBuffer.allocate(4)
        FfiConverterTypeBackendPreference.write(backend, buffer)
        check(buffer.array().contentEquals(byteArrayOf(0, 0, 0, (index + 1).toByte())))
        buffer.flip()
        check(FfiConverterTypeBackendPreference.read(buffer) == backend && !buffer.hasRemaining())
    }
    for (tag in listOf(0, 6, -1, Int.MAX_VALUE)) {
        val buffer = ByteBuffer.allocate(4).putInt(tag).flip()
        try {
            FfiConverterTypeBackendPreference.read(buffer)
            error("malformed production backend enum accepted")
        } catch (error: RuntimeException) {
            check(error.cause is IndexOutOfBoundsException && !buffer.hasRemaining())
        }
    }
    for (backend in listOf(BackendPreference.GPU, BackendPreference.METAL)) {
        for (direct in listOf(false, true)) {
            ModelLoader(Source.Bytes(bytes), EngineConfig(backend = backend)).use { loader ->
                try {
                    if (direct) loader.buildGenerative() else loader.build()
                    error("unavailable production backend accepted")
                } catch (error: uniffi.cera_ffi.LoadException.Assembly) {
                    check(error.backend == (if (backend == BackendPreference.GPU) "Gpu" else "Metal") && error.detail.isNotEmpty())
                }
                productionConsumed(loader)
            }
        }
    }
    val defaults = EngineConfig(backend = BackendPreference.CPU)
    check(defaults.contextSize == 4096uL && defaults.bundleRepo == null)
    check(defaults.draftModel == null && !defaults.gpuDepthformer)
    val sessionDefaults = SessionConfig()
    check(sessionDefaults.maxSeqLen == null && sessionDefaults.kvCompression == null)
    check(sessionDefaults.nKeep == 0u && sessionDefaults.seed == null)
    check(sessionDefaults.ubatchSize == 512u && !sessionDefaults.gpuDepthformer)
    val path = File(root, "model.gguf").absolutePath
    val sources =
        listOf(
            Source.Bytes(bytes),
            Source.Path(path),
            Source.Files(ModelFiles(path, null, null, null, null, emptyMap(), null, null)),
            Source.Parts(ModelParts(bytes, null, null, null, null, null, null, null)),
        )
    for (source in sources) {
        for (direct in listOf(false, true)) {
            val loader = ModelLoader(source, defaults)
            val handle = if (direct) null else loader.build()
            val model = if (direct) loader.buildGenerative() else checkNotNull(handle!!.asGenerative())
            productionConsumed(loader)
            val engine: CeraEngine = model.engine()
            model.engine().use { other -> check(enginesShareForProbe(engine, other)) }
            check(engine.contextSize() == 4096uL && engine.metadata().maxSeqLen == 64u)
            CeraEngine.fromBytes(bytes, defaults).use { independent ->
                check(!enginesShareForProbe(engine, independent))
            }
            val config =
                SessionConfig(
                    maxSeqLen = 8u,
                    kvCompression = KvCompression.F16,
                    seed = ULong.MAX_VALUE,
                    ubatchSize = 1u,
                    gpuDepthformer = true,
                )
            val session: Session = model.createSession(config)
            check(sessionConfigForProbe(session) == config)
            val sibling = engine.newSession(SessionConfig())
            loader.close()
            handle?.close()
            model.close()
            check(engine.contextSize() == 4096uL)
            check(engine.encodeText("ab") == listOf(0u, 1u))
            check(engine.decodeTokens(listOf(0u, 1u)) == "ab")
            sibling.appendTokens(listOf(1u))
            session.appendTokens(listOf(0u, 1u))
            engine.clearPrefixCache()
            check(session.position() == 2u && sibling.position() == 1u)
            engine.close()
            sibling.use {
                session.use {
                    val one = GenerateOpts(maxTokens = 1u, temperature = 0.7f, ignoreEos = true)
                    val first = session.generate(one)
                    check(first.tokens.size == 1 && session.position() == 3u)
                    val next = session.generate(GenerateOpts(maxTokens = 2u, temperature = 0.7f, ignoreEos = true))
                    check(next.tokens.size == 2 && session.position() == 5u)
                    check(sibling.position() == 1u)
                    session.reset()
                    check(session.position() == 0u)
                    try {
                        session.appendTokens(emptyList())
                        error("empty input accepted")
                    } catch (_: FfiException.EmptyInput) {
                    }
                    session.appendTokens(listOf(0u, 1u))
                    val sink = ProductionSink(session)
                    val streamed =
                        session.generateStreaming(
                            GenerateOpts(maxTokens = 3u, temperature = 0.7f, ignoreEos = true, flushEveryTokens = 1u),
                            sink,
                        )
                    check(sink.text.isNotEmpty() && sink.done.size == 1)
                    check(streamed.finishReason == FinishReason.Cancelled && streamed.tokensGenerated == 1u)
                    val position = session.position()
                    session.clearCancel()
                    session.generate(one)
                    check(session.position() == position + 1u)
                }
            }
        }
    }
    ModelLoader(Source.Bytes(bytes), defaults).use { loader ->
        loader.buildGenerative().use { model ->
            val modes =
                listOf(
                    null,
                    KvCompression.None,
                    KvCompression.F16,
                    KvCompression.TurboQuant(0uL, true, true),
                    KvCompression.TurboQuant(ULong.MAX_VALUE, true, false),
                    KvCompression.TurboQuant(42uL, false, true),
                )
            for (mode in modes) {
                val config = SessionConfig(8u, mode, 1u, 0uL, 0u, true)
                model.createSession(config).use { session ->
                    check(sessionConfigForProbe(session) == config.copy(kvCompression = mode ?: KvCompression.None))
                    session.appendTokens(listOf(0u, 1u))
                    check(session.generate(GenerateOpts(maxTokens = 1u, temperature = 0f, ignoreEos = true)).tokens.size == 1)
                }
            }
            model.createSession(SessionConfig(maxSeqLen = 1u)).use { capped ->
                capped.appendTokens(listOf(0u))
                try {
                    capped.appendTokens(listOf(1u))
                    error("session cap ignored")
                } catch (error: FfiException.ContextOverflow) {
                    check(error.maxSeqLen == 1u && error.by == 1u && capped.position() == 1u)
                }
            }
        }
    }
    return runProductionDefaults(bytes) + runProductionErrors(root, bytes) +
        listOf(
            "production-backend-transport",
            "production-defaults",
            "production-sources",
            "production-session",
            "production-shared-engine",
            "production-kv-config",
            "production-stream-cancel",
        )
}

private fun runProductionDefaults(bytes: ByteArray): List<String> {
    val empty = SamplingDefaults(null, null, null, null, null)
    val sampling = SamplingDefaults(0.37f, 0.71f, 7u, 0.13f, 1.23f)
    val profiles =
        mutableListOf<Triple<String, GenerationDefaults?, GenerationDefaults>>(
            Triple("absent", null, GenerationDefaults.Text(empty)),
            Triple("text-empty", GenerationDefaults.Text(empty), GenerationDefaults.Text(empty)),
        )
    for ((name, value) in listOf(
        "audio" to GenerationDefaults.Audio(sampling, 3u, 0.625f, 11u),
        "audio" to GenerationDefaults.Audio(empty, 0u, 0f, 0u),
        "audio" to GenerationDefaults.Audio(empty, UInt.MAX_VALUE, 1f, UInt.MAX_VALUE),
        "audio-empty" to GenerationDefaults.Audio(empty, null, null, null),
    )) {
        profiles += Triple(name, value, value)
    }
    for ((raw, canonical) in listOf(
        " { \"nested\" : [true, null, {\"x\":7}], \"label\": \"line\\ntext\" } " to
            "{\"label\":\"line\\ntext\",\"nested\":[true,null,{\"x\":7}]}",
        "[1, 2, null]" to "[1,2,null]",
        "42" to "42",
        "true" to "true",
        "\"text\"" to "\"text\"",
        "null" to "null",
    )) {
        profiles += Triple("other", GenerationDefaults.Other(raw), GenerationDefaults.Other(canonical))
    }
    for ((_, defaults, expected) in profiles) {
        for (typed in listOf(false, true)) {
            val parts = ModelParts(bytes, null, null, null, null, null, null, defaults)
            val loader = ModelLoader(Source.Parts(parts), EngineConfig(backend = BackendPreference.CPU))
            val handle = if (typed) null else loader.build()
            val model = if (typed) loader.buildGenerative() else checkNotNull(handle?.asGenerative())
            productionConsumed(loader)
            val observed = model.engine().use { productionDefaultsForProbe(it) }
            val session = model.createSession(SessionConfig(seed = 42uL))
            loader.close()
            handle?.close()
            model.close()
            check(observed == expected)
            session.use {
                it.appendTokens(listOf(0u, 1u))
                val output = it.generate(GenerateOpts(maxTokens = 3u, temperature = 0f, ignoreEos = true))
                check(output.tokens == listOf(0u, 1u, 0u) && it.position() == 5u)
            }
        }
    }
    for (raw in listOf("", "{", "null trailing", "{\"x\":NaN}")) {
        for (typed in listOf(false, true)) {
            val parts = ModelParts(bytes, null, null, null, null, null, null, GenerationDefaults.Other(raw))
            ModelLoader(Source.Parts(parts), EngineConfig(backend = BackendPreference.CPU)).use { loader ->
                try {
                    if (typed) loader.buildGenerative() else loader.build()
                    error("malformed defaults succeeded")
                } catch (error: LoadException.InvalidConfig) {
                    check(error.field == "generation_defaults.raw_json" && error.value == raw)
                    check(error.reason == "invalid_json" && error.detail.isNotEmpty())
                }
                productionConsumed(loader)
            }
        }
    }
    return profiles.map { "production-parts-defaults-${it.first}" }.distinct().sorted() + "production-parts-defaults-invalid-json"
}

private fun runProductionErrors(
    root: File,
    bytes: ByteArray,
): List<String> {
    val cases = mutableListOf<String>()
    val parts = ModelParts(bytes, null, null, null, null, null, null, null)
    for ((arch, kind) in listOf(
        "bert" to "Encoder",
        "modernbert" to "Encoder",
        "whisper" to "Whisper",
        "silero_vad" to "Vad",
        "kws" to "Hotword",
    )) {
        for (typed in listOf(false, true)) {
            ModelLoader(Source.Bytes(File(root, "$arch.gguf").readBytes()), EngineConfig(backend = BackendPreference.METAL)).use { bad ->
                try {
                    if (typed) bad.buildGenerative() else bad.build()
                    error("wrong kind succeeded")
                } catch (error: LoadException.KindMismatch) {
                    check(error.expected == "Generative" && error.actual == kind && error.architecture == arch) {
                        "$arch kind mismatch payload (typed=$typed)"
                    }
                }
                productionConsumed(bad)
            }
        }
        cases += "production-kind-$arch"
    }
    for (name in listOf("unknown", "malformed", "backend", "assembly", "inference")) {
        for (typed in listOf(false, true)) {
            val data =
                when (name) {
                    "unknown" -> File(root, "unknown.gguf").readBytes()
                    "assembly" -> File(root, "llama.gguf").readBytes()
                    "malformed" -> byteArrayOf(0, 1)
                    else -> bytes
                }
            val backend = if (name == "backend") BackendPreference.METAL else BackendPreference.CPU
            val source = if (name == "inference") Source.Parts(parts.copy(inferenceType = "future/unsupported")) else Source.Bytes(data)
            ModelLoader(source, EngineConfig(backend = backend)).use { bad ->
                try {
                    if (typed) bad.buildGenerative() else bad.build()
                    error("invalid load succeeded")
                } catch (error: LoadException.UnsupportedArchitecture) {
                    check(name == "unknown" && error.architecture == "future_probe")
                } catch (error: LoadException.Source) {
                    check(name == "malformed" && error.sourceKind == "bytes" && error.detail.isNotEmpty())
                } catch (error: LoadException.Assembly) {
                    check(name == "backend" || name == "assembly")
                    check(error.backend == (if (name == "backend") "Metal" else "Cpu") && error.detail.isNotEmpty())
                } catch (error: LoadException.UnsupportedInferenceType) {
                    check(name == "inference" && error.inferenceType == "future/unsupported")
                }
                productionConsumed(bad)
            }
        }
        cases += "production-$name"
    }
    return cases
}
